oidc-service-impl/src/main/java/io/mosip/esignet/services/ConsentHelperService.java

```diff
@@ -0,0 +1,224 @@
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toList());
+            transaction.setAcceptedClaims(acceptedClaims);
+            transaction.setPermittedScopes(permittedScopes);
+        }
+        cacheUtilService.setLinkedAuthenticatedTransaction(linkedAuthTransactionId,transaction);
+        if(consentAction.equals(ConsentAction.NOCAPTURE)){
+            cacheUtilService.setLinkedConsentedTransaction(linkedAuthTransactionId, transaction);
+            kafkaHelperService.publish(linkedAuthCodeTopicName, linkedAuthTransactionId);
+        }
+        LinkedKycAuthResponseV2 linkedKycAuthResponseV2 = new LinkedKycAuthResponseV2();
+        linkedKycAuthResponseV2.setLinkedTransactionId(linkedAuthTransactionId);
+        linkedKycAuthResponseV2.setConsentAction(consentAction);
+        return linkedKycAuthResponseV2;
+    }
+
+    public void addUserConsent(String transactionId, boolean linked) {
+        OIDCTransaction transaction;
+        if(!linked) {
+            transaction = cacheUtilService.getWebConsentedTransaction(transactionId);
+        } else {
+            transaction = cacheUtilService.getLinkedConsentedTransaction(transactionId);
+        }
+        if(transaction == null)
+            throw new InvalidTransactionException();
+        if(transaction.getConsentAction().equals(ConsentAction.CAPTURE)){
+            ConsentRequest consentRequest = new ConsentRequest();
+            consentRequest.setClientId(transaction.getClientId());
+            consentRequest.setPsuValue(transaction.getPartnerSpecificUserToken());
+            Claims claims = transaction.getRequestedClaims();
+            List<String> acceptedClaims = transaction.getAcceptedClaims();
+            setIsAcceptedBasedOnClaims(acceptedClaims,claims);
+            consentRequest.setClaims(claims);
+            List<String> permittedScopes = transaction.getPermittedScopes();
+            List<String> authorizeScope = transaction.getRequestedAuthorizeScopes();
+            Map<String, Boolean> authorizeScopes = permittedScopes != null ? permittedScopes.stream()
+                    .collect(Collectors.toMap(Function.identity(), authorizeScope::contains)) : Collections.emptyMap();
+            consentRequest.setAuthorizationScopes(authorizeScopes);
+            consentService.saveUserConsent(consentRequest);
+        }
+    }
+
+    private void setIsAcceptedBasedOnClaims(List<String> acceptedClaims, Claims requestedClaims) {
+        Map<String ,ClaimDetail> userInfo = requestedClaims.getUserinfo();
+        for (String acceptedClaim : acceptedClaims) {
+            if (userInfo.containsKey(acceptedClaim)) {
+                ClaimDetail claimDetail = userInfo.get(acceptedClaim);
+                if(claimDetail!= null){
+                    claimDetail.setAccepted(true);
+                } else {
+                    claimDetail = new ClaimDetail();
+                    claimDetail.setAccepted(true);
```

Comment:
Is this else condition possible ? Since accepted claims should be in the requested claim. Looks like only possible when someone is manipulating the request, in which case we should not take it

Reply:
this is possible for scenario where  a specific claim is sent as null (It is considered as optional), so when user accepts it we mark it as accepted

"userinfo": {
            "name": {
                "essential": false
            },
            "phone": null,
            "email": {
                "essential": true
            },
            "picture": {
                "essential": true
            },
            "gender": {
                "essential": true
            }
        },